flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SketchUtil.java

```diff
@@ -0,0 +1,182 @@
+   *
+   * @param numPartitions number of range partitions which equals to downstream operator parallelism
+   * @return reservoir size
+   */
+  static int determineCoordinatorReservoirSize(int numPartitions) {
+    int reservoirSize = numPartitions * COORDINATOR_TARGET_PARTITIONS_MULTIPLIER;
+
+    if (reservoirSize < COORDINATOR_MIN_RESERVOIR_SIZE) {
+      // adjust it up and still make reservoirSize divisible by numPartitions
+      int remainder = COORDINATOR_MIN_RESERVOIR_SIZE % numPartitions;
+      reservoirSize = COORDINATOR_MIN_RESERVOIR_SIZE + (numPartitions - remainder);
+    } else if (reservoirSize > COORDINATOR_MAX_RESERVOIR_SIZE) {
+      // adjust it down and still make reservoirSize divisible by numPartitions
+      int remainder = COORDINATOR_MAX_RESERVOIR_SIZE % numPartitions;
+      reservoirSize = COORDINATOR_MAX_RESERVOIR_SIZE - remainder;
+    }
+
+    return reservoirSize;
+  }
+
+  /**
+   * Determine the sampling reservoir size where operator subtasks collect data statistics.
+   *
+   * <p>Here are the heuristic rules
+   * <li>Target size is "coordinator reservoir size * over sampling ration (10) / operator
+   *     parallelism"
+   * <li>Min is 1K to achieve good accuracy while memory footprint is still relatively small
+   * <li>Max is 100K to cap the memory footprint on coordinator
+   *
+   * @param numPartitions number of range partitions which equals to downstream operator parallelism
+   * @param operatorParallelism data statistics operator parallelism
+   * @return reservoir size
+   */
+  static int determineOperatorReservoirSize(int operatorParallelism, int numPartitions) {
+    int coordinatorReservoirSize = determineCoordinatorReservoirSize(numPartitions);
+    int totalOperatorSamples = coordinatorReservoirSize * OPERATOR_OVER_SAMPLE_RATIO;
+    return (int) Math.ceil((double) totalOperatorSamples / operatorParallelism);
+  }
+
+  /**
+   * To understand how range bounds are used in range partitioning, heere is an example for human
+   * ages with 4 partitions: [15, 32, 60]. The 4 ranges would be
+   * <li>age <= 15
+   * <li>age > 15 && age <= 32
+   * <li>age >32 && age <= 60
+   * <li>age > 60
+   *
+   * @param numPartitions number of partitions which maps to downstream operator parallelism
+   * @param sketch aggregated reservoir sampling sketch
+   * @return list of range partition bounds. It should be a sorted list (ascending). Number of items
+   *     should be {@code numPartitions - 1}. if numPartitions is 1, return an empty list
+   */
+  static SortKey[] rangeBounds(
+      int numPartitions, Comparator<StructLike> comparator, ReservoirItemsSketch<SortKey> sketch) {
+    SortKey[] sortKeys = sketch.getSamples();
+    return determineBounds(Math.min(numPartitions, sortKeys.length), comparator, sortKeys);
+  }
+
+  /**
+   * This assumes the sort keys have equal weight, which is usually the case for high-cardinality
+   * scenarios (like device_id, user_id, uuid etc.).
+   */
```

Comment:
How do we handle cases where there is a high cardinality value and many small cardinality values? Especially for cases where the high cardinality value is high enough that multiple buckets should be used to write those?

Reply:
great question. I assume you meant high cardinality case but key weights are not evenly distributed (like hot key or some other non-even distributions like normal distributions).

Let me explain the implementation here and the Spark Partitioner. here, the array is sorted first. then samples are split evenly into range bounds. keys with higher weight should have more weights and the range should be smaller. Spark Partitioner calculate the weights of sampled keys and take that into account when computing the range bounds, which is similar to how the `MapRangePartitioner` works. I didn't want to convert the array of samples (which could be large) to a map with weights for every sampled keys. I plan to run some simulation to verify how well the current approach works. if it is bad, we can evaluate and switch the map weight calculation.

Reply:
Ok. I have missed the sampling part.
How are we make sure that our samples are representative?

Reply:
Oh... I see, the sketch implementation has a `getSamples` method. Let's run some tests. I understand your reasoning here now.

Reply:
I plan to add the simulation tests later in a separate PR, since this PR is very big already